rfd/0101-pod-rbac.md

```diff
@@ -0,0 +1,825 @@
+cluster and the pod name is included in the certificate's roles.
+
+When a `ResourceId.Kind` is `pod`, Teleport allows the user to access the Kubernetes
+cluster - even if only the `search_as_role` has RBAC rules to
+access it - and validates if the Pod name is allowed.
+
+#### Endpoints with POD's name
+
+For Endpoints whose path starts with `/api/v1/namespaces/{namespace}/pods/{name}`,
+the Teleport RBAC layer can extract the `{name}` from the request
+URI and validate if the pod name is in the user's allowed list. Teleport also computes
+at runtime the Kubernetes principals defined in the roles that match the cluster labels
+and pod name/namespace.
+
+Each time a user requests an endpoint prefixed with `/api/v1/namespaces/{namespace}/pods/{name}`,
+Teleport Kubernetes Service intercepts the request and extracts the `{name}`.
+After, it will validate if the `{name}` is included in the `kubernetes_resources`
+array or if it satisfies any wildcard pattern defined on it. If allowed, Teleport
+will forward the request to Kubernetes with the RBAC principals as Impersonation
+headers. Kubernetes will apply a second authorization layer using Kubernetes RBAC
+to validate the request. Finally, Teleport would forward the Kubernetes API
+response directly to the user without any modification.
+
+#### Special handlers
+
+When the Pod name is not accessible at request time, Teleport will have
+different behavior. Depending on the action the user is doing, Teleport will
+handle them separately.
+
+##### Listing Pods
+
+When listing Pods in the cluster or at a certain namespace, Teleport will receive
+the list of Pods from Kubernetes itself. Since Teleport depends on what Kubernetes
+RBAC allows users to access, Teleport would need to hijack the Kubernetes API
+response to exclude any Pod that is not included in the allowed list.
+
+To do that, Teleport will automatically forward the request to Kubernetes
+API for the following two endpoints:
+
+- `/api/v1/pods`
+  - `GET`: list or watch objects of kind Pods in every namespace.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `GET`: list or watch objects of kind Pods in namespace `{namespace}`.
+
+Once Teleport receives the response from Kubernetes API, it will unmarshal the
+response and exclude any Pod that is not in the allowed list. After excluding them,
+Teleport will return the filtered response to the user.
+
+##### Deleting Pods
+
+Kubernetes allows a user to delete every Pod in a certain namespace using a single
+call. The call does not include the name of the Pods to delete, so Teleport won't
+forward the request to Kubernetes. Instead, Teleport will list the user's Pods,
+exclude the ones that he shouldn't have access to, and issue a delete request for
+each one of the resulting Pod names.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `DELETE`: delete a collection of Pods in namespace `{namespace}`.
+
+Another option is to prevent this type of request in Teleport. This means that
```

Comment:
+1 for the "prevent this type of request in Teleport". 

In almost all cases if someone calls DELETE `{namespace_name}/pods` they mean "clear the given namespace", not "delete a few pods that I happen to have access too". So if the user does not have access to all of them, one can assume that the user made some kind of mistake, or doesn't fully understand what he's trying to do. Hence, the safer option is to just reject such operation.

Reply:
My opinion on this point is not clear.

The main points that make me think it's better to keep this feature are the following:
- from the user's perspective, the namespace only has the pods that he sees.
- delete pods support field selectors and labels which makes sense when filtering the objects to delete.
- I think Kube API has the same behavior, if you only have `delete` permissions for a single pod in the namespace and you try to delete all of them, only one will be deleted 

On the other hand, when we run `kubectl delete pods --all -n namespace`, kubectl does not use that endpoint, instead, it makes a list request and deletes pods 1 by 1.

Reply:
Got it. Keeping consistence with the Kube API is important too, so I don't have a strong opinion either, after all.